Proposal to use Souin as cache system (#17)
caddyserver/cache-handler: README.md

Caddy Module: http.handlers.cache
================================

> This is a distributed HTTP cache module for Caddy based on [Souin](https://github.com/darkweak/souin) cache.

Reviewer: Perhaps something like this? I think it makes the difference between the two clear.

I dropped the mention of "distributed" feature since that's mentioned below with other features of interest like RFC 7234 compliance.

It's not clear what the default is though, since you have both Badger (non-distributed, but can persist cache to disk?) and Olric (distributed and in-memory only), config docs can cover that or be reviewed on making it simpler (eg a `distributed` boolean instead of user needing to care about Badger or Olric providers and their configs, if they do they can change to Souin plugin).

It should be clear that distributed cache lacks on-disk persistence though (_perhaps that's not a feature `cache-handler` needs to accommodate for though?_).

```suggestion
The official Caddy HTTP cache module. This module embraces the Caddy ethos of keeping configuration simple with smart defaults out of the box!

Under the hood `cache-handler` uses the excellent HTTP caching solution [Souin](https://github.com/darkweak/souin). For advanced features like a REST API for purging cache entries or Surrogate Keys (aka YKey in Varnish), the [Souin Caddy module](https://github.com/darkweak/souin/tree/master/plugins/caddy) can be used directly when needed.
```

Author: The single `distributed` boolean option was an option but it was dropped to let the user decide to configure Olric by a configuration file or providing directly the config inside the Caddyfile to set max memory and other Olric options.

Author: I'm not sure about the `excellent` mention 😄

Author: > For advanced features like a REST API for purging cache entries or Surrogate Keys (aka YKey in Varnish), the [Souin Caddy module](https://github.com/darkweak/souin/tree/master/plugins/caddy) can be used directly when needed.
In fact the REST API is available inside the `cache-handler` but it's not enabled by default due to security reasons. And the `ykey` support may be possible (IMHO it should stay in both modules) but it won't be referenced in the `cache-handler` documentation.

## Features

>  * Supports most HTTP cache headers defined in [RFC 7234](https://httpwg.org/specs/rfc7234.html) (see the TODO section for limitations)

Reviewer: The TODO doesn't list any missing HTTP cache header support, does it no longer apply and should be removed?

```suggestion
 * Full support for HTTP cache headers defined in [RFC 7234](https://httpwg.org/specs/rfc7234.html).
```

or

```suggestion
 * [RFC 7234](https://httpwg.org/specs/rfc7234.html) compliant HTTP Cache.
```

Author: I think the second one is better.

>  * Sets [the `Cache-Status` HTTP Response Header](https://httpwg.org/http-extensions/draft-ietf-httpbis-cache-header.html)

Reviewer 2: Is this implemented?

Reviewer 3: Yes it is.

>  * REST API to be able to purge the cache and list stored keys.

Reviewer 2: ```suggestion
 * REST API to purge the cache and list stored resources
```

 * Builtin support for distributed cache.

> ## Example Configurations

Reviewer: Once decisions on feature set are decided, a snippet Caddyfile example in the README is usually better to present to users to get a feel for it.

Links can demonstrate more verbose setups/features.

>  * Embedded (or not) distributed cache can be used and a non-distributed one is supported too.

Reviewer 2: ```suggestion
 * Builtin support for distributed cache
```

You can find an example for the [Caddyfile](Caddyfile) or the [JSON file](configuration.json).  

> See the [Souin](https://github.com/darkweak/souin) configuration for the full configuration, and its associated [Caddyfile](https://github.com/darkweak/souin/blob/master/plugins/caddy/Caddyfile)  

Reviewer: Is it intended to remain compatible via config like this, or does the user need to switch to the Souin plugin to leverage the features being discussed as not supported in `cache-handler`?

Author: I think I should provide a minimal configuration for Caddy in the main Caddyfile (the one on at the root project directory).

> See the [Souin](https://github.com/darkweak/souin) configuration for the full configuration, and his associated [Caddyfile](https://github.com/darkweak/souin/plugins/caddy/Caddyfile)  

Reviewer 2: ```suggestion
See the [Souin](https://github.com/darkweak/souin) configuration for the full configuration, and its associated [Caddyfile](https://github.com/darkweak/souin/plugins/caddy/Caddyfile)  
```
